src/backend/access/transam/xlog.c

```diff
@@ -7201,6 +7221,14 @@ StartupXLOG(void)
 		ereport(PANIC,
 				(errmsg("invalid next transaction ID")));
 
+	if (ZenithRecoveryRequested)
+	{
+		if (wasShutdown)
+			checkPoint.oldestActiveXid = 0;
+		else if (!TransactionIdIsValid(checkPoint.oldestActiveXid))
+			checkPoint.oldestActiveXid = FirstNormalTransactionId;
```

Comment:
FirstNormalTransactionId seems completely bogus to me

Reply:
Do you think that this one is more correct:
```
		if (wasShutdown)
			checkPoint.oldestActiveXid = InvalidTransactionId;
		else if (!TransactionIdIsValid(checkPoint.oldestActiveXid))
			checkPoint.oldestActiveXid = GetOldestActiveTransactionId();
```

Reply:
I hope that with hack is not needed once `oldestActiveXid` is tracked by walingest.
But I still to keep this workaround.
Can you propose some alternative?
I tried to call `GetOldestActiveTransactionId()` but it can not be called while recovery is in progress.

Reply:
@knizhnik , can you please add a comment that this code branch covers unexpected situation.
and maybe also add some warning, so that we could spot such override in the logs.

Reply:
I have implemented now extraction of oldestActiveXid from "running XACTs" record any page server.
So it is expected that `checkPoint.oldestActiveXid` at replica should always contain valid XID (non zero).  Should we remove this check and replace it with assert? Frankly speaking I not sure that it makes code more reliable. Replacing 0 with `FirstNormalTransactionId`  is not 100% safe, because if XID is larger than 1^31, then we can loose some active transaction. But probability of it is very small and IMHO it is better then get assertion failure.